add notebooks and functional api (#714)
mosaicml/composer: docs/source/trainer/algorithms.rst

Under construction |:construction:|

> Composer has an awesome library of speedups that can be composed together
> to easily create efficient training recipes.

Reviewer: Hesitant to call our own library "awesome." Could probably call papers or methods "awesome" if we did it carefully.

The word "recipes" sounds normal to us but I don't know what it means when I put on my data scientist hat. Do I need training "recipes"? Is a "recipe" some abstraction in this library? I can kind of guess what it means but I'm not sure unless it's defined.

```suggestion
Composer has a collection of speedup methods that can be composed together
to easily create efficient training pipelines.
```

Reviewer: ^ although if "pipelines" suggests something more comprehensive to others (e.g., whole ETL pipeline), then "recipes" might be the least confusing phrasing. Or just "logic" or "scripts" or "programs". These might be safest.

    {% for name, data in metadata.items() %}
    {% if data.functional %}

>     :doc:`{{ data.class_name }}</method_cards/{{name}}>` | {{ data.tldr }} | :meth:`~composer.functional.{{ data.functional }}`

Reviewer: Doesn't create a link; maybe `:meth:` -> `:func:` will fix it.
```suggestion
    :doc:`{{ data.class_name }}</method_cards/{{name}}>` | {{ data.tldr }} | :func:`~composer.functional.{{ data.functional }}`
```
